### PR TITLE
added scope for active pieces to piece model

### PR DESCRIPTION
**Pull request by kyidkim**
Need to limit much of our existing work to apply only to active pieces. I added a scope to the pieces model - not sure if this will have the desired effect.

For example, it definitely needs to apply to the square_occupied? method - but that's in the game model. Will the scope in the piece model still apply?

Would appreciate any thoughts.

### CODE REVIEW

**Comment by cbohara** [comment]
I'm visualizing playing the game, and as I player I will only be trying to manipulate pieces that are on the board.  So the methods for moving pieces will be triggered when a player tries to move an active piece on the board.  A player will not be able to move captured pieces because they will no longer be on the board.  If that logic makes sense, I don't think it will be necessary to check if a piece is active before attempting to move it.

**Comment by kyidkim** [comment]
The issue I see is specifically with the method square_occupied in the game model.

When a piece is captured, it's coordinates are not removed (my mentor and I discussed this, and he recommended not setting them to nil). Currently, square_occupied has no means of checking that a piece is captured and should be excluded from its results.

We could just modify square_modified, but then we also have to stay aware in future development where similar bugs could pop up. Limiting the scope would be a better fix. I'm not sure if it works in the way I currently have it written though.

**Comment by jacobwgillespie** [comment]
So, we probably should do both (adjust square_occupied as well as check in the move method, etc.).

I'd recommend renaming the scope from `active_pieces` to just `active` because stuff like `Piece.active_pieces` or `game.pieces.active_pieces` is redundant (opinion).  And with your scope, adjusting `square_occupied` is pretty easy:

```ruby
def square_occupied?(x, y)
  pieces.active.where(x_position: x, y_position: y).any?
end
```

**Comment by kyidkim** [comment]
Updated with Jacob's recommendations

**Comment by cbohara** [comment]
LGTM!

File: app/models/game.rb [hunk 1]
```diff
@@ -49,20 +47,19 @@ def initialize_board!
...
-    pieces.where(x_position: x, y_position: y).any?
+    pieces.active.where(x_position: x, y_position: y).any?
   end
 
   def piece_data
-  data = Hash.new { |hash, key| hash[key] = {} }
-  pieces.each do |piece|
-    data[piece.x_position][piece.y_position]=piece
+    data = Hash.new { |hash, key| hash[key] = {} }
+    pieces.each do |piece|
```
Comment (jacobwgillespie) [inline app/models/game.rb]: Might also want `pieces.active.each` here as well...

**Comment by jacobwgillespie** [comment]
@kyidkim not sure if you've done a rebase before (ignore this if so), but if not this is a cheat-sheet version: https://github.com/team-rook/chess-app/pull/36#issuecomment-233424422

**Comment by kyidkim** [comment]
added active scope to the pieces data hash as well

**Comment by cbohara** [comment]
👍

**Pull request merged**